cpp/ql/src/semmle/code/cpp/exprs/Assignment.qll

```diff
@@ -168,11 +209,16 @@ class AssignPointerSubExpr extends AssignOperation, @assignpsubexpr {
 }
 
 /**
- * A C++ variable declaration in an expression where a condition is expected.
- * For example, on the `ConditionDeclExpr` in `if (bool c = x < y)`,
- * `getVariableAccess()` is an access to `c` (with possible casts),
- * `getVariable()` is the variable `c` (which has an initializer `x < y`), and
- * `getInitializingExpr()` is `x < y`.
+ * A C++ variable declaration inside the conditional expression of a `while` or `if`
+ * compound statement.  Declaring a variable this way narrows its lifetime and scope
+ * to be strictly the compound statement itself.  For example:
```

Comment:
I assume they can occur in a `for` loop as well?

Reply:
No, they cannot.  The initializer clause of a `for` can indeed contain declaration(s), but they are not conditional ones (i.e., compared against zero) as in `while` or `if`.

Reply:
What about:
```
for (i = 0; bool c = x < y; x++) ...
```

Reply:
No, that is not legal (though an interesting idea); the second clause must be an expression, not a declaration.

Reply:
It works for me: https://github.com/Semmle/ql/pull/1945